Add the option to configure memory ballast for Loki

SasSwart opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Name your PR as `<Feature Area>: Describe your change`.
  a. Do not end the title with punctuation. It will be added in the changelog.
  b. Start with an imperative verb. Example: Fix the latency between System A and System B.
  c. Use sentence case, not title case.
  d. Use a complete phrase or sentence. The PR title will appear in a changelog, so help other people understand what your change will be.
3. Rebase your PR if it gets out of sync with main
4. If changing the Helm chart, please ensure the chart version is increased per semantic versioning (https://semver.org)
-->

**What this PR does / why we need it**:
Adds the option to configure a memory ballast in order to enable reducing garbage collection when needed.
**Which issue(s) this PR fixes**:
Fixes #781

**Special notes for your reviewer**:
PR Based on reading [this](https://blog.twitch.tv/en/2019/04/10/go-memory-ballast-how-i-learnt-to-stop-worrying-and-love-the-heap/), and a comparable [PR in cortex](https://github.com/cortexproject/cortex/pull/1489/files).
**Checklist**
- [x] Documentation added
- [ ] Tests updated
- [x] Add an entry in the `CHANGELOG.md` about the changes.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/loki?pullRequest=5081) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/loki?pullRequest=5081) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/loki?pullRequest=5081) it.</sub>

dannykopping reviewed (approved):
LGTM, thanks for this @SasSwart!

SasSwart commented:
> LGTM, thanks for this @SasSwart! We'll definitely need some documentation about this, though. Do you mind adding that?

On it.

SasSwart commented:
@dannykopping Done. Added more clarity to flag description as well.

dannykopping commented:
> @dannykopping Done. Added more clarity to flag description as well.

Sorry, I should've been more clear. We have our docs at https://grafana.com/docs/loki/latest/configuration/ - which is linked to the [`docs`](https://github.com/grafana/loki/tree/main/docs) directory. The comments are worthwhile, for sure, so you can leave those as they are; I think we'll need an update to the [`yaml`](https://grafana.com/docs/loki/latest/configuration/#supported-contents-and-default-values-of-lokiyaml) config overview, perhaps plus a page in [`Operations`](https://grafana.com/docs/loki/latest/operations/) about how/when to use this ballast setting.

This is asking a lot so don't feel obligated; we can add that after merging the PR.

cyriltovena commented:
I agree I don't know when I should use it either

SasSwart commented:
Not sure how I missed that. Sorry y'all.

SasSwart commented:
While adding documentation, I noticed that I wasn't providing a tag for unmarshalling BallastBytes from the config file.

I've moved the BallastBytes value to the root config struct instead of the ConfigWrapper and provided the tag.

I also removed the option to set this as a Command Line argument in favor of the config file.
This seems to align better with the other configuration options.

dannykopping commented:
> While adding documentation, I noticed that I wasn't providing a tag for unmarshalling BallastBytes from the config file.
> 
> I've moved the BallastBytes value to the root config struct instead of the ConfigWrapper and provided the tag.
> 
> I also removed the option to set this as a Command Line argument in favor of the config file. This seems to align better with the other configuration options.

Thanks for this.

We generally allow for the setting of important config options via both YAML and CLI; would you mind restoring this one? It can be defined in `pkg/loki/loki.go` under `func (c *Config) RegisterFlags(f *flag.FlagSet)` like we do for `auth_enabled`, for example.

cyriltovena reviewed (approved):
LGTM

dannykopping commented:
Thanks @SasSwart :raised_hands: 
We'll wait for @KMiller-Grafana to approve the docs changes and then merge.

bboreham commented:
> reduce the amount of time spent resizing goroutine stack sizes.

May I ask what that relates to?  Normally I would say, as the code does, it's to reduce garbage-collection.

SasSwart commented:
Hey @bboreham 

That would be a wording mistake on my side. When opening the PR, I must've referenced the wrong section of [this issue](https://github.com/grafana/loki/issues/781). Thanks for spotting this.

I seem to have referred to [the top level comment](https://github.com/grafana/loki/issues/781#issue-470003524), whereas my contribution is more related to [this comment](https://github.com/grafana/loki/issues/781#issuecomment-527666824).

I'll update the PR description.

KMiller-Grafana commented on docs/sources/configuration/_index.md:
```diff
@@ -93,6 +93,12 @@ Pass the `-config.expand-env` flag at the command line to enable this way of set
 # if true. If false, the OrgID will always be set to "fake".
 [auth_enabled: <boolean> | default = true]
 
+# The amount of virtual memory to reserve as a ballast in order to optimise
```
Wording suggestion:
Change
> The amount of virtual memory to reserve as a ballast in order to optimise
to be
> The amount of virtual memory in bytes to reserve as ballast in order to optimize

KMiller-Grafana commented on docs/sources/operations/scalability.md:
```diff
@@ -22,3 +22,8 @@ In order to run with the Query Scheduler, the frontend needs to be passed the sc
 
 The query scheduler process itself can be started via the `-target=query-scheduler` option of the Loki Docker image. For instance, `docker run grafana/loki:latest -config.file=/cortex/config/cortex.yaml -target=query-scheduler -server.http-listen-port=8009 -server.grpc-listen-port=9009` starts the query scheduler listening on ports `8009` and `9009`.
+
+## Memory Ballast
+In compute constrained environments, garbage collection can become a significant factor. This can be optimised, at the expense of memory consumption, by configuring a memory ballast using the `ballast_bytes` configuration option.
+
+A larger memory ballast will mean that standard heap size volatility is less relatively significant, and therefore less likely to trigger garbage collection. This will result in lower compute overhead, but higher memory consumption, as the heap is cleaned less frequently.
```
We use sentence case for headers.  Here's my rewrite of this section of prose:

## Memory ballast

In compute-constrained environments, garbage collection can become a significant performance factor. Frequently-run garbage collection interferes with running the application by using CPU resources. The use of memory ballast can mitigate the issue. Memory ballast allocates extra, but unused virtual memory in order to inflate the quantity of live heap space. Garbage collection is triggered by the growth of heap space usage. The inflated quantity of heap space reduces the perceived growth, so garbage collection occurs less frequently.  

Configure memory ballast using the `ballast_bytes` configuration option.

SasSwart replied:
Thanks for the notes @KMiller-Grafana,

I'll prepare a follow up PR with these changes.

KMiller-Grafana reviewed (commented):
This PR was merged in before my doc suggestions were written.  Please consider using the suggestions in further revision.

Memory ballast is cool!
